ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java

```diff
@@ -1097,21 +1097,23 @@ private ServerErrorCode validateRequest(PartitionId partition, RequestOrResponse
     }
     if (!skipPartitionAndDiskAvailableCheck) {
       // 2. ensure the disk for the partition/replica is available
-      List<? extends ReplicaId> replicaIds = partition.getReplicaIds();
-      for (ReplicaId replica : replicaIds) {
-        if (replica.getDataNodeId().getHostname().equals(currentNode.getHostname())
-            && replica.getDataNodeId().getPort() == currentNode.getPort()) {
-          if (replica.getDiskId().getState() == HardwareState.UNAVAILABLE) {
-            metrics.diskUnavailableError.inc();
-            return ServerErrorCode.Disk_Unavailable;
-          }
-        }
+      ReplicaId localReplica = localPartitionToReplicaMap.get(partition);
+      if (localReplica != null && localReplica.getDiskId().getState() == HardwareState.UNAVAILABLE) {
+        metrics.diskUnavailableError.inc();
+        return ServerErrorCode.Disk_Unavailable;
       }
       // 3. check if partition exists on this node and that the store for this partition is available
       if (storageManager.getStore(partition) == null) {
-        if (partitionsInCurrentNode.contains(partition)) {
-          metrics.replicaUnavailableError.inc();
-          return ServerErrorCode.Replica_Unavailable;
+        if (localReplica != null) {
+          // check stores on the disk
+          if (!storageManager.isDiskAvailable(localReplica.getDiskId())) {
+            metrics.diskUnavailableError.inc();
+            localReplica.markDiskDown();
```

Comment:
is there any scenario where this needs to be marked up? It looks like, currently, this is a one way street. If a disk is marked down, it has no chance of coming back up again (even if a store is started). It seems to me that the disk state should be updated (by the disk manager?) if a store is started successfully on the disk.

Reply:
And need a test for that

Reply:
Added `markDiskUp` method. When store is successfully started, the disk on which the store sits will be marked as `AVAILABLE`

Reply:
Test is added in `StorageManagerTest`.